src/config.c

```diff
@@ -2691,7 +2691,7 @@ standardConfig configs[] = {
 
 void configCommand(client *c) {
     /* Only allow CONFIG GET while loading. */
-    if (server.loading && strcasecmp(c->argv[1]->ptr,"get")) {
+    if ((server.loading || server.async_loading) && strcasecmp(c->argv[1]->ptr,"get")) {
```

Comment:
Made the change, but need to confirm the places where it should be `(server.loading && !server.async_loading)`

From my search, this is what needs to be analyzed:
module.c
```
int RM_GetContextFlags(...)
    ...
    if (server.loading)
        flags |= REDISMODULE_CTX_FLAGS_LOADING;
```

multi.c
```
void execCommand(...)
    ...
    call(c,server.loading ? CMD_CALL_NONE : CMD_CALL_FULL);
```

rdb.c
```
void rdbReportError(...)
    ...
    if (!server.loading) {
        /* If we're in the context of a RESTORE command, just propagate the error. */
        /* log in VERBOSE, and return (don't exit). */
        serverLog(LL_VERBOSE, "%s", msg);
        return;
    }

...
serverLog(server.loading? LL_WARNING: LL_VERBOSE, "rdbLoadLzfStringObject failed allocating %llu bytes", (unsigned long long)clen);
...
serverLog(server.loading? LL_WARNING: LL_VERBOSE, "rdbLoadLzfStringObject failed allocating %llu bytes", (unsigned long long)len);
...
serverLog(server.loading? LL_WARNING: LL_VERBOSE, "rdbGenericLoadStringObject failed allocating %llu bytes", len);
...
serverLog(server.loading? LL_WARNING: LL_VERBOSE, "rdbGenericLoadStringObject failed allocating %llu bytes", len);
```

scripting.c
```
    /* Write commands are forbidden against read-only slaves, or if a
     * command marked as non-deterministic was already called in the context
     * of this script. */
    if (cmd->flags & CMD_WRITE) {
        int deny_write_type = writeCommandsDeniedByDiskError();
        if (server.lua_random_dirty && !server.lua_replicate_commands) {
            luaPushError(lua,
                "Write commands not allowed after non deterministic commands. Call redis.replicate_commands() at the start of your script in order to switch to single commands replication mode.");
            goto cleanup;
        } else if (server.masterhost && server.repl_slave_ro &&
                   !server.loading &&
                   !(server.lua_caller->flags & CLIENT_MASTER))
        {
            luaPushError(lua, shared.roslaveerr->ptr);
            goto cleanup;
        }
...

    /* If we reached the memory limit configured via maxmemory, commands that
     * could enlarge the memory usage are not allowed, but only if this is the
     * first write in the context of this script, otherwise we can't stop
     * in the middle. */
    if (server.maxmemory &&             /* Maxmemory is actually enabled. */
        !server.loading &&              /* Don't care about mem if loading. */
        !server.masterhost &&           /* Slave must execute the script. */
        server.lua_write_dirty == 0 &&  /* Script had no side effects so far. */
        server.lua_oom &&               /* Detected OOM when script start. */
        (cmd->flags & CMD_DENYOOM))
    {
        luaPushError(lua, shared.oomerr->ptr);
        goto cleanup;
    }
...
    /* If this is a Redis Cluster node, we need to make sure Lua is not
     * trying to access non-local keys, with the exception of commands
     * received from our master or when loading the AOF back in memory. */
    if (server.cluster_enabled && !server.loading &&
        !(server.lua_caller->flags & CLIENT_MASTER))
    {
        int error_code;
        /* Duplicate relevant flags in the lua client. */
```

Reply:
* module.c - i think we can keep the current code that depends only on `loading` so the module will know we're in loading state.
* multi.c - this is complicated, that code decides if `call` should put things in slowlog / stats and propagate to the AOF / replicas. so this may depend if the EXEC was received from the AOF file or another client, in either case, we don't want to propagate (and it'll also not allow any write commands), but for the slowlog and stats we wanna make that distinction.
* rdb.c - this may be similar to the above, i.e. a RESTORE command can fail differently when called from AOF vs a client. but actually, we don't expect a RESTORE command during loading on a read-only replica, so i think the code can stay as is (use `server.loading)
* scripting.c:
  * write commands - i don't currently understand how we could have got an EVAL command on a replica during loading before this PR. maybe that means a replica that's restarting from AOF? in that case, we wanna make a distinction again depending on the source of the command. commands that arrived from AOF should never fail, and ones that arrive from clients should.
  * oom - similar to the above. before this PR, EVAL commands from clients would be blocked during loading, and the check here was in order to never fail EVALs from AOF. we need to make a distinction depending on the client that triggered it.
  * cluster - same as the above.

bottom line, the main complication here is to detect the client while loading.
we could check `c->id == CLIENT_ID_AOF` instead of `server.loading`
or in the case of scripting.c check `server.lua_caller`.

@guybe7 @soloestoy i would love if you can validate these.

Reply:
in all the cases above (apart from module.c) we should check the client instead of server.loading

in module.c I suggest adding REDISMODULE_CTX_FLAGS_ASYNC_LOADING

Reply:
Ok about module.c we can add that. 
And indeed most other cases actually meant to check if the client is the aof client rather than look at the loading flag. 
However for rdb.c it's a bit more complicated, since the loading could be either RESTORE command (coming from the master client, an AOF or a normal client), but could also be an RDB file (not a command). 
So fat the loading flag check meant to distinguish between a restore command and an RDB file loading... 
So now we'll need two checks, first see if this is a command or an RDB file, and then check which client issued it. 
However, since we don't expect RESTORE commands during async loading, we can decide to keep the current code too..

On a second thought, I'd vote for the complicated one (being more explicit)

Reply:
@oranagra I don't quite get the c->id == CLIENT_ID_AOF check as solution. I'm missing something.
Let's take multi.c
```
if (c->id == CLIENT_ID_AOF)
    call(c,CMD_CALL_NONE);
else
    call(c,CMD_CALL_FULL);
```
For all situations:
- No SYNC or ASYNC loading: we're fine with this code, always call FULL.
- ASYNC loading, command is not from RDB or AOF: ok, will call FULL
- ASYNC loading, command from AOF: ok, will call NONE
- ASYNC loading, command from RDB (replication, for example): ?
- SYNC loading: what if it's loading from RDB, CLIENT_ID_AOF won't cover it and we will end up calling FULL, right?

Reply:
I think that what you're missing is that RDB file doesn't contain commands, just raw data. So the only ways for `call` to be called during loading is that it's either from the AOF fake client or real client connection.

Reply:
Changes done, but the complicated way for rdb.c would get quite complicated as we don't have the client there.
Do you believe it worth passing around for the explicitness?

Reply:
In rdb.c we can use `server.current_client`
If it's NULL then we're processing an RDB file (or a preamble AOF). 
If it's non-null we can check the ID to see if it's the AOF client. 
If it non null and not the aof client, then it's a restore command.
Like other places, no need to look at the `loading` flag at all. 
Let's wrap that in some macro at the top and document that.